File: O/901-online-stock-span/main.go
package main

type StockSpanner struct {
	NS []int
	IS []int
	C  int
}

func Constructor() StockSpanner {
	return StockSpanner{NS: []int{}, IS: []int{}}
}

func (this *StockSpanner) Next(price int) int {
	this.C++
	if len(this.NS) == 0 {
		this.NS = append(this.NS, price)
		this.IS = append(this.IS, this.C)
		return 1
	}

	idx := len(this.NS) - 1
	for idx >= 0 && this.NS[idx] <= price {
		idx--
	}

	v := this.C
	if idx >= 0 {
		v = this.C - this.IS[idx]
	}

	this.NS = this.NS[0 : idx+1]
	this.IS = this.IS[0 : idx+1]
	this.NS = append(this.NS, price)
	this.IS = append(this.IS, this.C)
	return v
}

File: O/901-online-stock-span/main.rs
struct Solution {}

struct StockSpanner {
  b: Vec<i32>,
  i: Vec<i32>,
  c: i32,
}

impl StockSpanner {
  fn new() -> Self {
    StockSpanner {
      b: Vec::new(),
      i: Vec::new(),
      c: 0,
    }
  }

  fn next(&mut self, price: i32) -> i32 {
    self.c += 1;
    if self.b.len() == 0 || self.b[self.b.len() - 1] > price {
      self.b.push(price);
      self.i.push(self.c - 1);
      return 1;
    }

    while self.b.len() > 0 && self.b[self.b.len() - 1] <= price {
      self.b.pop();
      self.i.pop();
    }

    if self.b.len() == 0 {
      self.b.push(price);
      self.i.push(self.c - 1);
      return self.c;
    }

    self.b.push(price);
    self.i.push(self.c - 1);
    self.c - 1 - self.i[self.i.len() - 2];
  }
}
